vortex-expr/src/is_null.rs

```diff
@@ -0,0 +1,173 @@
+use std::any::Any;
+use std::fmt::Display;
+use std::ops::Not;
+use std::sync::Arc;
+
+use vortex_array::arrays::{BoolArray, ConstantArray};
+use vortex_array::{Array, ArrayRef, IntoArray};
+use vortex_dtype::{DType, Nullability};
+use vortex_error::VortexResult;
+use vortex_mask::Mask;
+
+use crate::{ExprRef, VortexExpr};
+
+#[derive(Debug, Eq, Hash)]
+#[allow(clippy::derived_hash_with_manual_eq)]
+pub struct IsNull {
+    child: ExprRef,
+}
+
+impl IsNull {
+    pub fn new_expr(child: ExprRef) -> ExprRef {
+        Arc::new(Self { child })
+    }
+}
+
+impl PartialEq for IsNull {
+    fn eq(&self, other: &Self) -> bool {
+        self.child.eq(&other.child)
+    }
+}
+
+impl Display for IsNull {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "is_null({})", self.child)
+    }
+}
+
+#[cfg(feature = "proto")]
+pub(crate) mod proto {
+    use vortex_error::{VortexResult, vortex_bail};
+    use vortex_proto::expr::kind;
+    use vortex_proto::expr::kind::Kind;
+
+    use crate::is_null::IsNull;
+    use crate::{ExprDeserialize, ExprRef, ExprSerializable, Id};
+
+    pub(crate) struct IsNullSerde;
+
+    impl Id for IsNullSerde {
+        fn id(&self) -> &'static str {
+            "is_null"
+        }
+    }
+
+    impl ExprDeserialize for IsNullSerde {
+        fn deserialize(&self, kind: &Kind, children: Vec<ExprRef>) -> VortexResult<ExprRef> {
+            let Kind::IsNull(kind::IsNull {}) = kind else {
+                vortex_bail!("wrong kind {:?}, want is_null", kind)
+            };
+
+            Ok(IsNull::new_expr(children[0].clone()))
+        }
+    }
+
+    impl ExprSerializable for IsNull {
+        fn id(&self) -> &'static str {
+            IsNullSerde.id()
+        }
+
+        fn serialize_kind(&self) -> VortexResult<Kind> {
+            Ok(Kind::IsNull(kind::IsNull {}))
+        }
+    }
+}
+
+impl VortexExpr for IsNull {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn unchecked_evaluate(&self, batch: &dyn Array) -> VortexResult<ArrayRef> {
+        match batch.validity_mask()? {
+            Mask::AllTrue(len) => Ok(ConstantArray::new(false, len).into_array()),
+            Mask::AllFalse(len) => Ok(ConstantArray::new(true, len).into_array()),
+            Mask::Values(mask) => Ok(BoolArray::from(mask.boolean_buffer().not()).into_array()),
+        }
+    }
+
+    fn children(&self) -> Vec<&ExprRef> {
+        vec![&self.child]
+    }
+
+    fn replacing_children(self: Arc<Self>, children: Vec<ExprRef>) -> ExprRef {
+        assert_eq!(children.len(), 1);
+        Self::new_expr(children[0].clone())
```

Comment:
I would `.pop()` here and then expect the option you get since you have an assert before. You could also remove the assert and just do `Self::new_expr(children.pop().vortex_expect(..))` which should be the same